runners/google-cloud-dataflow-java/src/main/java/org/apache/beam/runners/dataflow/DataflowMetrics.java

```diff
@@ -0,0 +1,210 @@
+        "namespace", metricUpdate.getName().getContext().get("namespace"),
+        "step", metricUpdate.getName().getContext().get("step")
+    );
+  }
+
+  /**
+   * Check whether a {@link com.google.api.services.dataflow.model.MetricUpdate} is a tentative
+   * update or not.
+   * @param metricUpdate
+   * @return true if update is tentative, false otherwise
+   */
+  private boolean isMetricTentative(
+      com.google.api.services.dataflow.model.MetricUpdate metricUpdate) {
+    return (metricUpdate.getName().getContext().containsKey("tentative")
+        && Objects.equal(metricUpdate.getName().getContext().get("tentative"), "true"));
+  }
+
+  /**
+   * Take a list of metric updates coming from the Dataflow service, and format it into a
+   * Metrics API MetricQueryResults instance.
+   * @param metricUpdates
+   * @return a populated MetricQueryResults object.
+   */
+  private MetricQueryResults populateMetricQueryResults(
+      List<com.google.api.services.dataflow.model.MetricUpdate> metricUpdates,
+      MetricsFilter filter) {
+    // Separate metric updates by name and by tentative/committed.
+    HashMap<ImmutableMap<String, String>, com.google.api.services.dataflow.model.MetricUpdate>
+        tentativeByName = new HashMap<>();
+    HashMap<ImmutableMap<String, String>, com.google.api.services.dataflow.model.MetricUpdate>
+        commitedByName = new HashMap<>();
+    HashSet<ImmutableMap<String, String>> metricHashKeys = new HashSet<>();
+
+    // If the Context of the metric update does not have a namespace, then these are not
+    // actual metrics counters.
+    for (com.google.api.services.dataflow.model.MetricUpdate update : metricUpdates) {
+      if (Objects.equal(update.getName().getOrigin(), "user") && isMetricTentative(update)
+          && update.getName().getContext().containsKey("namespace")) {
+        tentativeByName.put(metricHashKey(update), update);
+        metricHashKeys.add(metricHashKey(update));
+      } else if (Objects.equal(update.getName().getOrigin(), "user")
+          && update.getName().getContext().containsKey("namespace")
+          && !isMetricTentative(update)) {
+        commitedByName.put(metricHashKey(update), update);
+        metricHashKeys.add(metricHashKey(update));
+      }
+    }
+    // Create the lists with the metric result information.
+    ImmutableList.Builder<MetricResult<Long>> counterResults = ImmutableList.builder();
+    ImmutableList.Builder<MetricResult<DistributionResult>> distributionResults =
+        ImmutableList.builder();
+    for (ImmutableMap<String, String> metricKey : metricHashKeys) {
+      String metricName = metricKey.get("name");
+      if (metricName.endsWith("[MIN]") || metricName.endsWith("[MAX]")
+          || metricName.endsWith("[MEAN]") || metricName.endsWith("[SUM]")) {
+        // Skip distribution metrics, as these are not yet properly supported.
+        LOG.warn("Distribution metrics are not yet supported. You can see them in the Dataflow"
```

Comment:
Yes, that's right. There's currently a problem with distributions with structured names (it's my bad. I just discovered it). Perhaps I'll add that change after I fix that.